src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AttachedCollections/NuGet/AssetsFileTopLevelDependenciesCollectionSourceProvider.cs

```diff
@@ -0,0 +1,119 @@
+
+        protected abstract bool TryUpdateItem(TItem item, AssetsFileTarget targetData, AssetsFileTargetLibrary library);
+
+        protected override bool TryCreateCollectionSource(
+            IVsHierarchyItem hierarchyItem,
+            string flagsString,
+            string? target,
+            IRelationProvider relationProvider,
+            [NotNullWhen(returnValue: true)] out AggregateRelationCollectionSource? containsCollectionSource)
+        {
+            if (!TryGetIdentity(flagsString, out TIdentity identity))
+            {
+                containsCollectionSource = null;
+                return false;
+            }
+
+            UnconfiguredProject? unconfiguredProject = hierarchyItem.HierarchyIdentity.Hierarchy.AsUnconfiguredProject();
+
+            // Find the data source
+            IAssetsFileDependenciesDataSource? dataSource = unconfiguredProject?.Services.ExportProvider.GetExportedValueOrDefault<IAssetsFileDependenciesDataSource>();
+
+            if (unconfiguredProject == null || dataSource == null)
+            {
+                containsCollectionSource = null;
+                return false;
+            }
+
+            // Items for top-level dependencies do not appear in the tree directly. They "shadow" hierarchy items
+            // for purposes of bridging between the hierarchy items (top-level dependencies) and attached items (transitive
+            // dependencies).
+            TItem? item = null;
+
+            var collectionSource = new AggregateRelationCollectionSource(hierarchyItem);
+            AggregateContainsRelationCollection? collection = null;
+
+            IDisposable link = dataSource.SourceBlock.LinkToAsyncAction(
+                async versionedValue =>
+                {
+                    AssetsFileDependenciesSnapshot snapshot = versionedValue.Value;
+                    if (snapshot.TryGetTarget(target, out AssetsFileTarget? targetData))
+                    {
+                        if (TryGetLibrary(targetData, identity, out AssetsFileTargetLibrary? library))
+                        {
+                            if (item == null)
+                            {
+                                // This is the first update
+                                item = CreateItem(targetData, library);
+                                if (AggregateContainsRelationCollection.TryCreate(item, relationProvider, out collection))
+                                {
+                                    await _joinableTaskContext.Factory.SwitchToMainThreadAsync();
+                                    collectionSource.SetCollection(collection);
+                                }
+                            }
+                            else if (TryUpdateItem(item, targetData, library) && collection != null)
+                            {
+                                await _joinableTaskContext.Factory.SwitchToMainThreadAsync();
```

Comment:
I think you want to use IProjectThreadingService from IProjectService here. This is effectively part of the project, and you want to make sure you end up going through CPS's JTF collection.